contracts/contracts/strategies/ConvexOUSDMetaStrategy.sol

```diff
@@ -0,0 +1,176 @@
+        ICurvePool curvePool = ICurvePool(platformAddress);
+
+        uint256 threePoolLpBalance = IERC20(pTokenAddress).balanceOf(address(this));
+        uint256 curve3PoolVirtualPrice = curvePool.get_virtual_price();
+        uint256 threePoolLpDollarValue = threePoolLpBalance.mulTruncate(
+            curve3PoolVirtualPrice
+        );
+
+        uint256 ousdToAdd = toPositive(
+            int256(
+                metapool.balances(crvCoinIndex).mulTruncate(
+                    curve3PoolVirtualPrice
+                )
+            ) -
+                int256(metapool.balances(mainCoinIndex)) +
+                int256(threePoolLpDollarValue)
+        );
+
+        /* Add so much OUSD so that the pool ends up being balanced. And at minimum
+         * add twice as much OUSD as 3poolLP and at maximum at twice as 
+         * much OUSD. 
+         */
+        ousdToAdd = Math.max(ousdToAdd, threePoolLpDollarValue);
+        ousdToAdd = Math.min(ousdToAdd, threePoolLpDollarValue * 2);
+
+        /* Mint so much ousd that the dollar value of 3CRVLP in the pool and OUSD will be equal after
+         * deployment of liquidity. In cases where pool is heavier in OUSD before the deposit strategy mints
+         * less OUSD and gets less metapoolLP and less rewards. And when pool is heavier in 3CRV strategy mints
+         * more OUSD and gets more metapoolLP and more rewards.
+         *
+         * In both cases metapool ends up being balanced and there should be no incentive to execute arbitrage trade.
+         */
+        if (ousdToAdd > 0) {
+            IVault(vaultAddress).mintForStrategy(ousdToAdd);
+        }
+
+        uint256 ousdBalance = metapoolMainToken.balanceOf(address(this));
+        uint256[2] memory _amounts = [ousdBalance, threePoolLpBalance];
+
+        uint256 metapoolVirtualPrice = metapool.get_virtual_price();
+        /**
+         * First convert all the deposited tokens to dollar values,
+         * then divide by virtual price to convert to metapool LP tokens
+         * and apply the max slippage
+         */
+        uint256 minReceived = (ousdBalance + threePoolLpDollarValue)
+            .divPrecisely(metapoolVirtualPrice)
+            .mulTruncate(uint256(1e18) - maxSlippage);
+
+        // slither-disable-next-line unused-return
+        metapool.add_liquidity(_amounts, minReceived);
+
+        uint256 metapoolLp = metapoolErc20.balanceOf(address(this));
+
+        bool success = IConvexDeposits(cvxDepositorAddress).deposit(
+            cvxDepositorPTokenId,
+            metapoolLp,
+            true // Deposit with staking
+        );
+
+        require(success, "Failed to deposit to Convex");
+    }
+
+    /**
+     * Withdraw the specified amount of tokens from the gauge. And use all the resulting tokens
+     * to remove liquidity from metapool
+     * @param num3CrvTokens Number of Convex LP tokens to remove from gauge
+     */
+    function _lpWithdraw(uint256 num3CrvTokens) internal override {
```

Comment:
I think this withdraw can still get a bit simpler.


```python
needed = (num3CrvTokens + 1) * metapool.totalSupply()  // metapool.balances(1)

convex.withdrawAndUnwrap(needed, true);
ousd_metapool.remove_liquidity(needed, [0,0])
ousd.burn(ousd.balanceOf(address(this))
```


Since there are no fees when removing with remove_liquidity, we don't have to worry about fees.

Since remove_liquidity removes both sides in the proportion between pool balances, lp tokens burnt, and total supply, we can perfectly calculate how much LP tokens we need to use (outside of solidity rounding error.). Add a one and we don't have to worry about rounding. (Check this though.)

Because we are withdrawing both coins, we will have OUSD unless we are withdrawing less than a couple raw LP tokens. We might as well just keep the code clean and always burn.

Reply:
Ran this code 
```python
#############
## setup metapool
from metastrategy import *
threepool_lp.approve(ousd_metapool.address, 0, {"from": me})
threepool_lp.approve(ousd_metapool.address, 1e50, {"from": me})
ousd_metapool.add_liquidity([0, 1e24], 0, {"from":me}) #add 1mio liquidity

num3CrvTokens = 1e23
needed = (num3CrvTokens + 1) * ousd_metapool.totalSupply() // ousd_metapool.balances(1)
user_3crv_balance = threepool_lp.balanceOf(me)
ousd_metapool.remove_liquidity(needed, [0,0], {"from":me})
user_3crv_diff = threepool_lp.balanceOf(me) - user_3crv_balance
user_3crv_diff
```
and it outputs 99999999999999990439768

Good news is that like you say fee is not included in liquidity removal.  Bad news is there seems to be slightly more calculation error than expected. I am also slightly worried that different pool implementations might have different error rates. I would suggest to add a ful extra coin to required 3poolRemoval
```
needed = (num3CrvTokens + 1e18) * ousd_metapool.totalSupply() // ousd_metapool.balances(1)
```
This should give plenty of room for rounding errors and shouldn't cause any serious problems with funds being left behind on contracts

Reply:
Oh ignore the rambling regarding different pool implementations. Looks like an extra 1e7 of 3crvLP should make up for the calculation error. For safe measure rather adding 1e10 of extra 3crv LP

Reply:
thanks @DanielVF great suggestions. Have further simplified the code see this commit: https://github.com/OriginProtocol/origin-dollar/pull/1002/commits/61f506beffcd3e13a3cc0586dd60e310b2e5a03c

Reply:
Re-ran the above test code with depositing 1e26 tokens and withdrawing 1e25. The output of `user_3crv_diff` was: 
10000000000000000554802370

Interesting that the calculation sometimes overshoots and sometimes undershoots.

Reply:
I wonder if 1e10 is enough of a margin for error.

Reply:
Another calculation approach, using a higher precision division:

```python
k = int(1e36) * total_supply // balance
diff = (balance) * k - (balance - amount - 1) * k
to_burn = diff // int(1e36)
```

Haven't thought it all the way through though. Works with what I've played with.